keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1089 @@
+in the parameter type. The validation can set bounds that are specific to the
+driver.
+
+```
+const GroupName = "dra.e2e.example.com"
+const KindName = "ClaimParameters"
+const Version = "v1alpha1"
+
+type ClaimParameters struct {
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	metav1.TypeMeta   `json:",inline"`
+
+	Spec ClaimParameterSpec `json:"spec"`
+}
+
+type ClaimParameterSpec struct {
+	// +kubebuilder:default=1
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=16
+	NumberOfFrobnicators int64 `json:"numberOfFrobnicators,omitempty"`
+
+	// Additional configuration parameters, in this case
+	// some environment variables for the container in which
+	// the Frobnicators are going to be used.
+	Env map[string]string `json:"env,omitempty"`
+}
+```
+
+The following template is used to generate a
+`counter.dra.config.k8s.io/Parameters` description from a
+`dra.e2e.example.com/ClaimParameters` object:
+
+```
+kind: Parameters
+apiVersion: counter.dra.config.k8s.io/v1alpha1
+Required: {{object.spec.numberOfFrobnicators}}
+```
+
+The content in the `{{ ... }}` placeholder is a
+[CEL](https://kubernetes.io/docs/reference/using-api/cel/) expression. The
+`object` variable provides access to the fields in the vendor CRD. Note that
+this particular example driver doesn't use the `Selector` feature. Users never
+get to see it either.
+
+### Notes/Constraints/Caveats
+
+For this proposal to work, kube-scheduler must have permission to read, list
+and watch the custom claim parameter resources. This is necessary because it is not known in
+advance which CRDs will be used by DRA drivers. When deploying a driver,
+suitable role bindings for
+[`system:kube-scheduler`](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#core-component-roles)
+have to be added which grant read, watch and list permissions for the vendor
+CRD.
+
+This may change at runtime as classes get added or removed. To avoid repeated
+GET operations for those CRDs, dynamic informers will have to be used that get
+started and stopped as needed.
+
+## Design Details
+
+### ResourceClass extension
+
+A new, optional field in ResourceClass enables numeric parameters for claims
+using this class:
+
+```
+type ResourceClass struct {
+	...
+
+	// If (and only if) allocation of claims using this class is handled
+	// via numeric parameters, then NumericParameters must list all claim
+	// parameter types that are allowed for claims.
+	NumericParameters []NumericParameterType
+}
+
+type NumericParameterType struct {
+	// APIVersion is the group and version for the resource being referenced
+	// as claim parameter.
+	// +optional
+	APIVersion string
+
+	// Kind is the type of resource being referenced. This is the same
+	// value as in the parameter object's metadata.
+	Kind string
+
+	// CELTemplate gets expanded with the claim parameter object as input.
+	// The version of that object will be as specified by APIVersion.
+	CELTemplate string
```

Comment:
@thockin [was concerned](https://github.com/kubernetes/enhancements/pull/4384#issuecomment-1918109611) that admins will not get this right:
> I don't buy this - what will really happen 99% of the time is that the vendor will provides N suggested classes with the mappings. Admins who change that will likely break their clusters.

I agree that this is a risk. Admins have to set up the class exactly as instructed by their vendor. What are the alternatives?

- Drop templating, rely on DRA CRD controller to convert and publish as status: [complicated, has its own challenges](https://github.com/kubernetes/enhancements/pull/4384#issuecomment-1918663267)
- Reference a config map that comes as part of the driver installation: simple?
- Reference an entirely new type  that comes as part of the driver installation: more complex, not sure whether a separate type just to store a string is justified.

Admins can still get it wrong, but at least what they can get wrong is smaller when storing the template elsewhere.

Reply:
One advantage of putting the CELTemplate into its own type (whether it is ResourceClass or something new) is that the apiserver can parse and validate the template.

Reply:
@thockin: another concern that @johnbelamaric brought up: is templating another no-go from an API and architecture perspective? Does it depend on how we do the templating?

There's [a thread](https://github.com/kubernetes/enhancements/pull/4384#discussion_r1471870744) about "expression types" or string templates, but both share the same problem (can fail during evaluation).

Reply:
Reading, and I don't want to over-state before really comprehending, but... templates make things hard to reason-about and are my least-preferred path for building something humans can use.

Reply:
Agreed. Let's continue in https://github.com/kubernetes/enhancements/pull/4384/files#r1474085092.